chain/pruned_block_dispatcher_test.go

```diff
@@ -0,0 +1,593 @@
+package chain
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/btcjson"
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/peer"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightninglabs/neutrino/query"
+	"github.com/lightningnetwork/lnd/ticker"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	addrCounter int32 = 0 // Increased atomically.
+)
+
+func nextAddr() string {
+	port := atomic.AddInt32(&addrCounter, 1)
+	return fmt.Sprintf("10.0.0.1:%d", port)
+}
+
+// conn mocks a network connection by implementing the net.Conn interface. It is
+// used to test peer connection without actually opening a network connection.
+type conn struct {
+	io.Reader
+	io.Writer
+	io.Closer
+	localAddr  string
+	remoteAddr string
+}
+
+func (c conn) LocalAddr() net.Addr {
+	return &addr{"tcp", c.localAddr}
+}
+func (c conn) RemoteAddr() net.Addr {
+	return &addr{"tcp", c.remoteAddr}
+}
+func (c conn) SetDeadline(t time.Time) error      { return nil }
+func (c conn) SetReadDeadline(t time.Time) error  { return nil }
+func (c conn) SetWriteDeadline(t time.Time) error { return nil }
+
+// addr mocks a network address.
+type addr struct {
+	net, address string
+}
+
+func (m addr) Network() string { return m.net }
+func (m addr) String() string  { return m.address }
+
+// pipe turns two mock connections into a full-duplex connection similar to
+// net.Pipe to allow pipe's with (fake) addresses.
+func pipe(c1, c2 *conn) (*conn, *conn) {
+	r1, w1 := io.Pipe()
+	r2, w2 := io.Pipe()
+
+	c1.Writer = w1
+	c1.Closer = w1
+	c2.Reader = r1
+	c1.Reader = r2
+	c2.Writer = w2
+	c2.Closer = w2
+
+	return c1, c2
+}
+
+// genBlockChain generates a test chain with the given number of blocks.
+func genBlockChain(numBlocks int) ([]*chainhash.Hash, map[chainhash.Hash]*wire.MsgBlock) {
+	prevHash := new(chainhash.Hash)
+	hashes := make([]*chainhash.Hash, numBlocks)
+	blocks := make(map[chainhash.Hash]*wire.MsgBlock, numBlocks)
+	for i := 0; i < numBlocks; i++ {
+		header := wire.NewBlockHeader(0, prevHash, prevHash, 0, 0)
+		block := wire.NewMsgBlock(header)
+
+		blockHash := block.BlockHash()
+		hashes[i] = &blockHash
+		blocks[blockHash] = block
+
+		prevHash = &blockHash
+	}
+
+	return hashes, blocks
+}
+
+// networkBlockTestHarness is a harness used to facilitate the testing of the
+// PrunedBlockDispatcher.
+type networkBlockTestHarness struct {
+	t *testing.T
+
+	hashes []*chainhash.Hash
+	blocks map[chainhash.Hash]*wire.MsgBlock
+
+	peerMtx     sync.Mutex
+	peers       map[string]*peer.Peer
+	localConns  map[string]net.Conn // Connections to peers.
+	remoteConns map[string]net.Conn // Connections from peers.
+
+	dialedPeer    chan struct{}
+	queriedPeer   chan struct{}
+	blocksQueried map[chainhash.Hash]int
+
+	shouldReply uint32 // 0 == true, 1 == false
+
+	*PrunedBlockDispatcher
```

Comment:
I think it would be clearer if this had its own field perhaps?

Reply:
Done.